scipy/optimize/_hessian_update_strategy.py

```diff
@@ -187,15 +187,35 @@ def update(self, delta_x, delta_grad):
             return
         if self.first_iteration:
             # Get user specific scale
-            if self.init_scale == "auto":
+            if isinstance(self.init_scale, str) and self.init_scale == "auto":
                 scale = self._auto_scale(delta_x, delta_grad)
             else:
-                scale = float(self.init_scale)
+                scale = self.init_scale
+
+            # Check for complex: numpy will silently cast a complex array to
+            # a real one but not so for scalar as it raises a TypeError.
+            # Checking here brings a consistent behavior.
+            if np.size(scale) == 1:
+                # to account for the legacy behavior having the exact same cast
+                scale = float(scale)
+            if np.iscomplexobj(scale):
```

Comment:
I meant this part. This wasn't here and it will block the scalar route. We only need `if isinstance(scale, ndarray)` route before the `float()`, this here introduces new behavior. Previously it was all passing through float so only new options should be redirected anything else would still go through float.

Reply:
Sidenote: it may be correct to do an elif instead of an if, but since a Python float surely never is a complex object, this will never trigger the second if so it won't change the discussion

For the rest, I rather disagree:

> this here introduces new behavior

Yes, that is the _very goal_ of the PR! Adding more possibilities _will_ change the behavior. Some input in the past will now trigger different behavior. E.g. an np.array threw an error, now it won't and can be used.

However, that is fine. It is never guaranteed or documented (!) that an np.array throws an error. It was illegal and the consequences have never been defined.

However, we can catch the _reasonable_ cases such as if a complex value is given, one _may_ relies on the fact that an error is thrown. Giving in an array and getting a TypeError I would not count to it though.

_why not simply `isinstance`_: This again puts very tight constraints on the objects. What about other array-like objects? Tensors? No need to restrict us to arrays. Anything that has a shape and can handle a multiplication should be allowed.

So my take is: this will change the behavior. However, to restore the "reasonable" legacy behavior, we can say that the behavior of the function for size-1 objects (which is a large definition) remains the same. For anything else, it will change.

What do you think? Or rather, which input will now change the behavior in the past _and is worth to be conserved_?

Reply:
Unfortunately there are many packages depending on us and we are tied up to keeping everything intact unless a very compelling need arises. New features are fine but old behavior needs to stay exactly as it is. If I passed 3+0.j that was passing now it is not. That's a no go regardless whether this is good or not.

Anything that was undocumented was kind of hanging in the balance but here you are clearly blocking that is not really needed to be blocked. That is the problem. There is no benefit gained from this block. My earlier comment was about since you removed the float() initially everything was passing in and numpy would upcast everything in the case of an accidental complex number. These are different issues. If you put down one guard you have to replace it somewhere else.

Reply:
> Unfortunately there are many packages depending on us and we are tied up to keeping everything intact unless a very compelling need arises. New features are fine but old behavior needs to stay exactly as it is. 

Of course, and I fully agree. But there is still the question on what is "reasonable": there are version numbers on packages for a certain reason; this does not mean you should break features all the time but really weird behavior, well yes. That may just changes. Or there should be a difference between feature (stays the same, no question), behavior (can be reasonable to keep the same) and "undefined behavior" basically. The latter is what I would say is by passing an array in now.

> If I passed 3+0.j that was passing now it is not. That's a no go regardless whether this is good or not.

No, it was never passing. And it will not fail because of the `np.iscomplexobj` but because of the `float` that was already there before.

To be precise, it trows `TypeError: can't convert complex to float`

It always an error. Having the check for complex and throwing an error for larger objects just _aligns_ the new additional feature behavior with what was there.

Reply:
So to be clear: whenever a `np.size == 1` object is passed, the behavior is *exactly* the same.

This includes a whole array of things, also complex numbers.

(_okay: under the assumption that `matrix * float` is the same as `matrix * np.array(float)`, which is the case_)

P.S: NOT under the assumption that the float is a float, that we know

Reply:
I have never used complex numbers in Python but I would be very surprised if any nonlinear optimizer in  `scipy.optimize` worked with them. In principle, I agree with @jonas-eschle that if anyone relied on supplying a complex number here was using a bug not a feature. Still, backwards compatibility ..

Could we perform the complex number check only for the newly added array possibility? That would avoid changing old behaviour.

Reply:
We already did this here: the legacy way was to cast to a float, and that's exactly what we're doing now as well. This _will_ fail for a complex value, so the `np.iscomplex` will never be reached and only applies to the new arrays. But I just changed the code a bit anyways, just to be more specific with shapes etc, and that naturally included an `elif`

The only non-backwards compatible change is for arrays, that a) won't fail anymore (that's the feature we're adding) or b) will fail for other reasons (which is now the case as I've added the additional checks).

(just as an overall sidenote, this PR could nearly be considered a bug fix instead of a new feature: it seemed quite surprising that the init value, the value to initialize an _array_, was _cast_ to a float, and passing a numpy array as initialization _was not possible_. With all the general numpy array compatibility and some freedom for broadcasts...)